### PR TITLE
[TIMELINE] Create a dynamic image for social medias

### PR DESCRIPTION
**Pull request by Remg**

### CODE REVIEW

File: src/Repository/Timeline/MeasureRepository.php [hunk 1]
```diff
@@ -21,4 +21,19 @@ public function findOneByTitle(string $title): ?Measure
             ->getOneOrNullResult()
         ;
     }
+
+    public function countMeasuresByStatus(string $status): int
+    {
+        $qb = $this
```
Comment (HeahDude) [inline src/Repository/Timeline/MeasureRepository.php]: You can put the `return` here.

File: src/Timeline/TimelineImageFactory.php [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        return $this->measureRepository->countMeasuresByStatus($status);
+    }
+
+    private function translate(string $id, string $locale): string
+    {
+        return $this->translator->trans($id, [], 'messages', $locale);
+    }
+
+    private function createImageFromPng(string $path)
```
Comment (HeahDude) [inline src/Timeline/TimelineImageFactory.php]: Static, same for others below.

File: src/Timeline/TimelineImageFactory.php [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    private function getMeasureCounts(): array
+    {
+        foreach (array_keys(self::TEXTS_COORDINATES) as $measureStatus) {
+            $counts[$measureStatus] = $this->countMeasuresByStatus($measureStatus);
+        }
+
+        return $counts ?? [];
+    }
+
+    private function countMeasuresByStatus(string $status): int
```
Comment (HeahDude) [inline src/Timeline/TimelineImageFactory.php]: This method has no benefit.

File: src/Timeline/TimelineImageFactory.php [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    public function createImage(string $locale): string
+    {
+        $counts = $this->getMeasureCounts();
+
+        $image = $this->createImageFromPng($this->getImageTemplatePath($locale));
+
+        $this->drawTexts($image, $counts, $locale);
+        $this->drawChart($image, $counts);
+
+        $this->saveImage($image, $imagePath = $this->getOutputImagePath());
```
Comment (HeahDude) [inline src/Timeline/TimelineImageFactory.php]: No need to pass the second arg, inline the method call in `saveImage` and return the output path from there.

File: src/Timeline/TimelineImageFactory.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        $this->translator = $translator;
+        $this->cacheDirectory = $cacheDirectory;
+        $this->webDirectory = $webDirectory;
+    }
+
+    public function createImage(string $locale): string
+    {
+        $counts = $this->getMeasureCounts();
+
+        $image = $this->createImageFromPng($this->getImageTemplatePath($locale));
```
Comment (thomasbisignani) [inline src/Timeline/TimelineImageFactory.php]: What happens if an error occurs here ?

Comment (Remg) [reply src/Timeline/TimelineImageFactory.php]: `If he dies, he dies.`

File: src/Timeline/TimelineImageFactory.php [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        $color1 = $this->createColor($image, 255, 234, 0);
+        $color2 = $this->createColor($image, 189, 182, 81);
+        $background = $this->createColor($image, 113, 129, 255);
+
+        $this->drawChartArc($image, self::CHART['outerRadius'], $color1end, $color2end, $color2);
+        $this->drawChartArc($image, self::CHART['outerRadius'], $color1start, $color1end, $color1);
+        $this->drawChartArc($image, self::CHART['innerRadius'], 0, 360, $background);
+    }
+
+    private function getOutputImagePath(): string
```
Comment (HeahDude) [inline src/Timeline/TimelineImageFactory.php]: Can be inlined.

File: src/Timeline/TimelineImageFactory.php [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        return sprintf('%s/%s', $this->cacheDirectory, self::OUTPUT_IMAGE_NAME);
+    }
+
+    private function getImageTemplatePath(string $locale): string
```
Comment (HeahDude) [inline src/Timeline/TimelineImageFactory.php]: Same.

File: src/Timeline/TimelineImageFactory.php [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        $this->cacheDirectory = $cacheDirectory;
+        $this->webDirectory = $webDirectory;
+    }
+
+    public function createImage(string $locale): string
+    {
+        $counts = $this->getMeasureCounts();
+
+        if (!$image = $this->createImageFromPng($this->getImageTemplatePath($locale))) {
+            throw new \InvalidArgumentException("Image template doest not exist for locale \"$locale\".");
```
Comment (thomasbisignani) [inline src/Timeline/TimelineImageFactory.php]: `does not`

**Pull request merged**